libs/pilight/core/common.c

```diff
@@ -1019,3 +1024,161 @@ int file_get_contents(char *file, char **content) {
 	fclose(fp);
 	return 0;
 }
+
+int checkdnsrr(const char *domain, const char *type)
+{
+	logprintf(LOG_STACK, "%s(%s,%s)", __FUNCTION__, domain ? domain : "(NULL)", type ? type : "(NULL)");
+	return 0;	// NYI.
+}
+
+#if defined(__FreeBSD__) || defined(_WIN32)
+#define use_regexp 0
+#else
+#define use_regexp 1
+#endif
+
+// Do NOT write the next as range A-Za-z0-9 because the string may be used without RE.
+#define AtoZ_atoz_0to9 "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
+#define white_spaces " \t"
+
+// Next email_local_part_valid_chars is defined in a way that it can be used without RE
+// and within [] of a posix extended regular expression. Therefore the dash comes first
+// because at later position it would mean range (and it can't be escaped because
+// a \ inside [] is literal - this is different from perl RE).
+#define email_local_part_valid_chars "-" AtoZ_atoz_0to9 ".!#$%&'*+\\/=?^_`{|}~"
+
+#if use_regexp
+// The RE to check email addr is compiled only once and used commonly.
+
+static regex_t check_email_common_RE = { 0 };
+static const regex_t * volatile check_email_compiled_RE = NULL;
+
+static void check_email_free_RE() {
+	if(check_email_compiled_RE != NULL) {	// protect against unexpected calls.
+		regfree(&check_email_common_RE);
+		check_email_compiled_RE = NULL;
+	}
+}
+#endif
+
+/**
+ * Check if an email address is valid.
+ * @param char* addr The email address to check. Leading+trailing white spaces are accepted.
+ * @param bool allow_lists If true, addr is allowed to be a comma separated list of email addresses.
+ * @param bool check_domain_can_mail If true, check if domain exists in DNS system and accepts email.
+ * @return int Values < 0 indicate error. Values >= 0 indicate success.
+ * Where -1: syntax error, -2: DNS check for a domain failed, -8: severe internal error.
+ */
+int check_email_addr(const char *addr, int allow_lists, int check_domain_can_mail) {
+	if(addr == NULL) {
+		return -1;
+	}
+#if use_regexp
+	const regex_t *RE = NULL;
+	int reti = 0;
+
+	static const char email_pattern[] =
+		"^[" email_local_part_valid_chars "]+"
```

Comment:
Not shure what you mean.

The defines are not outsite of the check-fxn to ease validation of email addresses anywhere else (would this be my intention, I defined it in common.h). It's just because a ```#define ``` is not bound to function scope. Its not a big difference to write a #define in the function body or before. If you meant this, I can put the #defines inside. It just wouldn't change that the #defines remain for the rest of the file (unless #undef'd).

If you meant to have a (static and/or const) variable in the function body that holds the string snippets, the answer is: I didn't do that here because I make use of letting the compiler concatenate string literals (at compile time) out of the string snippets that are ```#defined```. Strings in variables must be concatenated at run time and would cause me to write code that does it (and it would be done on each call despite its always the same "constant" result).